desc/equilibrium/equilibrium.py

```diff
@@ -836,60 +888,100 @@ def compute(
         if data is None:
             data = {}
 
-        # To avoid the issue of using the wrong grid for surface and volume averages,
-        # we first figure out what needed qtys are flux functions or volume integrals
-        # and compute those first on a full grid
         p = "desc.equilibrium.equilibrium.Equilibrium"
-        # If the user wants to compute x which depends on y which in turn depends on z,
-        # and they pass in y already computed in data, then we shouldn't need to compute
-        # z at all.
-        deps = list(
+        # TODO:
+        #  If the user wants to compute x which depends on y which in turn depends on z,
+        #  and they pass in y already in data, then we shouldn't need to compute z at
+        #  all. To resolve this we need to compute dependencies recursively, so that
+        #  priming data with just the first order dependencies will avoid computing
+        #  unnecessary dependencies. For now just subtract out y.
+        deps = (
             set(get_data_deps(names, obj=p, has_axis=grid.axis.size) + names)
-            - data.keys()  # subtract out y if already computed
+            - data.keys()
         )
-        # TODO: replace this logic with `grid_type` from data_index
-        dep0d = [
-            dep
-            for dep in deps
-            if (data_index[p][dep]["coordinates"] == "") and (dep not in data)
-        ]
-        dep1dr = [
+
+        def need_src(name):
+            # Need to compute these on grid that is paired to the source grid, since
+            # the compute logic assume input data is evaluated on those coordinates.
+            # We exclude these from the depXdx sets below since the grids we will
+            # use to compute those dependencies are coordinate-blind.
+            return bool(data_index[p][name]["source_grid_requirement"])
+
+        need_src_deps = _grow_seeds(set(filter(need_src, deps)), deps)
+
+        dep0d = {dep for dep in deps if self.is_0d(dep) and dep not in need_src_deps}
+        # Unless user asks, don't try to recompute stuff which are only dependencies
+        # of dep0d. Example, need R0. R0 <- A <- A(z) computable on dep0d grid.
+        # But A(z) in dep1dz and attempt to recompute on dep1dz grid will error
+        # without unique zeta idx defined, which mapped coordinate grids lack.
+        dep0d_deps = set(get_data_deps(dep0d, obj=p, has_axis=grid.axis.size))
```

Comment:
@rahulgaur104 this is the issue you are running, except instead of `R0` you need `a`, and even after computing `a` correctly on a quadrature grid, `master` tries to compute a dependency of `a`, in particular `A(z)` on a `LinearGrid` which errors when doing field line stuff because the field line grids don't have `unique_zeta_idx` defined.

Reply:
Okay, maybe we can merge it fast so I can fix ballooning because right now, ballooning calculations fail.
@dpanici  can you review this PR this week so we can merge it quickly.

Reply:
```sh
# for security make new local branch that is copy of branch you want to merge
git switch branch_i_want_to_merge_this_PR
git checkout -b new_branch
git merge --squash fieldline_compute
```
The squash merge will merge the branch as one commit.  May be useful if you would like to unmerge with
```sh
git stash
```

Reply:
Well there is another problem now. There is a change in this PR that allows the evaluation to work but throws an error when I optimize with the BallooningObjective. Note, that the current ballooning PR (not merged with this PR) works fine for optimization but throws an error during evaluation. 
So the error during optimization is definitely coming from this PR. You can find more details here: https://github.com/PlasmaControl/DESC/blob/370168ef6aa67934b645a472193fa1478a916f5d/docs/notebooks/tutorials/ideal_ballooning_stability.ipynb

Reply:
I pulled your fork `rg/adjoint_ballooning2` and ran the optimization under the cell block "Optimizing for ballooning stability". I was unable to reproduce the objective build error you linked. My machine is fedora linux with jax 0.4.28 on cpu.

This PR won't affect optimization; we don't call `eq.compute` in objectives. If your bug is reproducible, perhaps some other pull request since May 8th has affected numerics. Check #1022?

Reply:
Yeah, I think my jax version was 0.4.23 which causes it to throw that that error. Updating to jax 0.4.28 gets rid of the error.